lib/action_cable/connection/hijacking_protection.rb

```diff
@@ -0,0 +1,97 @@
+require 'active_support/security_utils'
+
+module ActionCable
+  module Connection
+    # Action Cable connection are protected from Cross-Site WebSocket Hijacking
+    # (CSWSH) by using the same CSRF token protecting Rails' controllers.
+    # When starting a Websocket connection the Action Cable javascript will
+    # send along the CSRF token used to protect controllers. The token is
+    # stored as a random string in the session, to which an attacker does not
+    # have access. When a Websocket connection request reaches your
+    # application, Action Cable verifies the received token with the token in
+    # the session.
+    #
+    # CSWSH protection is turned on with the <tt>protect_from_hijacking</tt>
+    # method.
+    #
+    # module ApplicationCable
+    #   class Connection < ActionCable::Connection::Base
+    #     protect_from_hijacking, if: -> { current_user.guest? }
+    #   end
+    # end
+    #
+    # Make sure that every page that requires starting a Websocket connection
+    # includes in it's head the CSRF. This is accomplised by including
+    # <tt>csrf_meta_tags</tt> in the HTML +head+. Action Cable needs to have
+    # access to the meta tags so make sure that the <tt>csrf_meta_tags</tt> is
+    # above the <tt>javascript_link_tag</tt> or the javascript / coffescript
+    # creating the cable is executed on a domready event.
+    #
+    # Learn more about CSWSH attacks and securing your application:
+    # {Cross-Site WebSocket Hijacking}[https://www.christian-schneider.net/CrossSiteWebSocketHijacking.html].
+    # {Ruby on Rails Security Guide}[http://guides.rubyonrails.org/security.html].
+    module HijackingProtection
+      extend ActiveSupport::Concern
+
+      AUTHENTICITY_TOKEN_LENGTH = 32
+
+      module ClassMethods
+        def protect_from_hijacking(options = {})
+          options = options.reverse_merge(prepend: true)
+          after_connect :validate_connection, options
+        end
+      end
+
+
+      protected
+      # This is the actual method verifying the token
+      def validate_connection
+        reject_unauthorized_connection "Invalid CSWSH/CSRF token" unless valid_authentication_token?(request.params["token"])
+      end
+
+      def valid_authentication_token?(encoded_masked_token)
+        if encoded_masked_token.nil? || encoded_masked_token.empty? || !encoded_masked_token.is_a?(String)
```

Comment:
Can't we use `blank?` instead of `nil? || `empty?`?

We also don't care about returning singletons, so we could simplify as:

```ruby
return unless encoded_masked_token.present? && encoded_masked_token.is_a?(String)
```